### PR TITLE
removing deprecated workflow API and appropriate changes in Workflow …

### PR DESCRIPTION
**Pull request by gsps1**
…Driver to not use Workflow Action to run programs.

Most of the code is removal of deprecated API, except changes in WorkflowDriver and ProgramWorkflowRunnerFactory, to run the programs directly instead of using WorkflowAction wrapper. As WorkflowAction is also deprecated and removed.

Build : https://builds.cask.co/browse/CDAP-DUT6187

### CODE REVIEW

**Review by anwar6953 — APPROVED** [review]

**Comment by anwar6953** [comment]
LGTM

**Comment by chtyim** [comment]
please resolve conflicts.

File: cdap-app-fabric-tests/src/test/java/co/cask/cdap/WorkflowAppWithScopedParameters.java [hunk 1]
```diff
@@ -206,21 +206,11 @@ public void destroy() {
...
-      WorkflowNodeState nodeState = nodeStates.get(ONE_MR);
-      Preconditions.checkArgument(ONE_MR.equals(nodeState.getNodeId()));
-      Preconditions.checkArgument(nodeState.getRunId() != null);
-      Preconditions.checkArgument(NodeStatus.COMPLETED == nodeState.getNodeStatus());
-
-      nodeState = nodeStates.get(ONE_SPARK);
-      Preconditions.checkArgument(ONE_SPARK.equals(nodeState.getNodeId()));
-      Preconditions.checkArgument(nodeState.getRunId() != null);
-      Preconditions.checkArgument(NodeStatus.COMPLETED == nodeState.getNodeStatus());
+      // no-op
```
Comment (chtyim) [inline cdap-app-fabric-tests/src/test/java/co/cask/cdap/WorkflowAppWithScopedParameters.java:213]: Why this becomes no-op? I thought it is part of the test?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowContext.java [hunk 2]
```diff
@@ -50,27 +49,25 @@
...
+  BasicWorkflowContext(WorkflowSpecification workflowSpec,
                        WorkflowToken token, Program program, ProgramOptions programOptions, CConfiguration cConf,
                        MetricsCollectionService metricsCollectionService,
                        DatasetFramework datasetFramework, TransactionSystemClient txClient,
                        DiscoveryServiceClient discoveryServiceClient, Map<String, WorkflowNodeState> nodeStates,
                        @Nullable PluginInstantiator pluginInstantiator,
                        SecureStore secureStore, SecureStoreManager secureStoreManager,
                        MessagingService messagingService, @Nullable ConditionSpecification conditionSpecification) {
-    super(program, programOptions, cConf, (spec == null) ? new HashSet<String>() : spec.getDatasets(),
+    super(program, programOptions, cConf, new HashSet<String>(),
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowContext.java]: You can just use `new HashSet<>()`. Java 8 can infer the type.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java [hunk 3]
```diff
@@ -332,7 +317,24 @@ private void executeAction(WorkflowActionNode node, WorkflowToken token) throws
...
+          // this should not happen, since null is only passed in from WorkflowDriver, only when calling configure
+          if (programWorkflowRunner == null) {
+            throw new UnsupportedOperationException("Operation not allowed.");
+          }
+
+          Runnable programRunner = programWorkflowRunner.create(programName);
+          LOG.info("Starting {} Program '{}' in workflow", prettyProgramType, programName);
+          programRunner.run();
+
+          // TODO (terence) : Put something back to context.
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java]: I don't think we need this anymore. Please have it removed.

File: cdap-app-fabric-tests/src/test/java/co/cask/cdap/WorkflowAppWithScopedParameters.java (hunk 1 above)
Comment (gsps1) [reply cdap-app-fabric-tests/src/test/java/co/cask/cdap/WorkflowAppWithScopedParameters.java:213]: `WorkflowContext` was available from `AbstractWorkflowAction` which had the node states. however now `CustomActionContext` will only have the state of the custom action available via `getState`. 

As for the test, workflow node states are available to `destroy` method and we perform checks there already.

Comment (chtyim) [reply cdap-app-fabric-tests/src/test/java/co/cask/cdap/WorkflowAppWithScopedParameters.java:213]: I see

**Comment by gsps1** [comment]
@chtyim any further comments ?

**Comment by gsps1** [comment]
build green, merging.

**Pull request merged**